File: medium/0347-top-k-frequent-elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function (nums, k) {
  // Count occurences using a hashmap
  const occurences = new Map();
  nums.forEach((num) => {
    if (occurences.has(num)) {
      occurences.set(num, occurences.get(num) + 1);
    } else {
      occurences.set(num, 1);
    }
  });

  // Sort hashmap in descending order of values/occurences
  const sortedMap = [...occurences].sort((a, b) => b[1] - a[1]);

  const res = [];
  for (let i = 0; i < k; i++) {
    res.push(sortedMap[i][0]);
  }
  return res;

  // One liner of above block of code
  // return sortedMap.slice(0, k).map(pair => pair[0]);
};

// Time complexity: O(nlogn)
// Space complexity: O(n)

File: medium/0347-top-k-frequent-elements.py
from typing import List
from collections import Counter


# Time complexity: O(nlogn)
# Space complexity: O(n)
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # hashmap with key: num and val: freq of num
        # iterate over nums and fill hashmap
        # sort hashmap by descending order of values
        # return k most frequent elements as a list
        count = Counter(nums)
        sortedCount = dict(
            sorted(count.items(), key=lambda item: item[1], reverse=True)
        )
        return list(sortedCount.keys())[:k]
